File: dasgupta/ch2/merge_sort.py
#!/usr/bin/env python3

def merge_sort(nums: list) -> list:
    """Implement merge sort algorithm.

    Args:
        nums: List of comparable elements to be sorted

    Returns:
        A new list:  sorted nums in ascending order

    Example:
        >>> merge_sort([5, 2, 8, 1, 9])
        [1, 2, 5, 8, 9]
        >>> merge_sort([])
        []
        >>> merge_sort([1])
        [1]
    """

    if len(nums) <= 1:
        return nums

    mid = len(nums) // 2
    left = nums[:mid]
    right = nums[mid:]

    # Conquer: recursively sort each half
    left_sorted = merge_sort(left)
    right_sorted = merge_sort(right)

    # Combine: merge the sorted halves
    return merge(left_sorted, right_sorted)


def merge(left: list, right: list) -> list:
    """Merge two **sorted** lists into a single sorted list.
    Runtime:  O(n)

    Args:
        left: First sorted list
        right: Second sorted list

    Returns:
        A **new** sorted list containing all elements from both input lists

    Example:
        >>> merge([1, 3, 5], [2, 4, 6])
        [1, 2, 3, 4, 5, 6]
    """

    result = []
    i = j = 0

    # Compare elements from left and right, adding the smaller one to result
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Add any remaining elements from left
    while i < len(left):
        result.append(left[i])
        i += 1

    # Add any remaining elements from right
    while j < len(right):
        result.append(right[j])
        j += 1

    return result

File: dasgupta/ch2/cpp/merge-sort/merge-sort.cpp
#include "merge-sort.hpp"

using std::vector;

vector<int>
merge(const vector<int>& left, const vector<int>& right) {

    vector<int> result;
    result.reserve(left.size() + right.size());

    size_t i = 0; // left-half index
    size_t j = 0; // right-half index

    // Select minimum of two, and build result vector<int>
    while (i < left.size() && j < right.size())
        left[i] <= right[j] ?
            result.push_back(left[i++]) :
            result.push_back(right[j++]);

    // Handle the remainings.
    while (i < left.size()) 
        result.push_back(left[i++]);
    while (j < right.size()) 
        result.push_back(right[j++]);

    return result;    // will use move-semantics?
}

vector<int> merge_sort(const vector<int>& nums) {

    // Special Case.
    if (nums.size() <= 1) {
        return nums;
    }

    size_t mid = nums.size() / 2;

    // left = nums[:mid]
    std::vector<int> left;
    for (auto it = nums.begin(); it != nums.begin() + mid; ++it)
        left.push_back(*it);

    // right = nums[mid+1:]
    std::vector<int> right;
    for (auto it = nums.begin() + mid; it != nums.end(); ++it)
        right.push_back(*it);
    
    vector<int> left_sorted = merge_sort(left);
    vector<int> right_sorted = merge_sort(right);

    // Recursive Merge and Return.
    return merge(left_sorted, right_sorted);
}
